### PR TITLE
[Storage] `sky storage create` CLI

### PR DESCRIPTION
**Pull request by choochutran**
Implemented `sky storage create` cli command that allows users to create a Sky storage from their terminal. Included examples, edited documentation, and wrote a test in test_smoke.py to test creation and subsequent deletion of sky storage object. 

Tested: Created and verified creation of sky storage using cli command in test_sky_storage_cli test in test_smoke.py.

### CODE REVIEW

File: sky/cli.py [hunk 1]
```diff
@@ -2635,6 +2635,32 @@ def storage_delete(names: Tuple[str], all: bool):  # pylint: disable=redefined-b
         sky.storage_delete(name)
 
 
+@storage.command('create', cls=_DocumentedCodeCommand)
+@click.argument('name', required=False, type=str, nargs=1)
+@click.argument('source', required=False, type=str, nargs=1)
+@click.argument('stores', required=False, type=str, nargs=-1)
```
Comment (michaelzhiluo) [inline sky/cli.py:2641]: By default, it should be `S3`

File: sky/cli.py [hunk 2]
```diff
@@ -2635,6 +2635,32 @@ def storage_delete(names: Tuple[str], all: bool):  # pylint: disable=redefined-b
         sky.storage_delete(name)
 
 
+@storage.command('create', cls=_DocumentedCodeCommand)
+@click.argument('name', required=False, type=str, nargs=1)
+@click.argument('source', required=False, type=str, nargs=1)
```
Comment (michaelzhiluo) [inline sky/cli.py:2640]: For all arguments, add a `help` description to it. (For reference look at how other CLI click arguments are created)

File: sky/cli.py [hunk 3]
```diff
@@ -2635,6 +2635,32 @@ def storage_delete(names: Tuple[str], all: bool):  # pylint: disable=redefined-b
...
+            local path or a cloud URI (s3://, gs://, etc.). Local paths do not
+            need to be absolute.
+          stores: Optional; Specify pre-initialized stores (S3 or GCS).
+
+    Examples:
+
+    .. code-block:: bash
+
+        # Create a storage object with GCS and AWS stores
+        sky storage create imagenet S3://bucket-name GCS AWS
```
[the remark this answers is not in the document]
Comment (michaelzhiluo) [reply sky/cli.py:2657]: In addition, does this command semantically work? I would assume you would need `--source` and `--stores`.

File: sky/core.py [hunk 4]
```diff
@@ -757,3 +757,21 @@ def storage_delete(name: str) -> None:
...
+def storage_create(name: Optional[str] = None,
+                   source: Optional[str] = None,
+                   stores: Tuple[str] = None):
+    """"Create a storage.
+
+    """
+    storage = data.Storage(name, source)
+
+    stores = [] if stores is None else stores
+    stores = [store_type.upper() for store_type in stores]
```
Comment (michaelzhiluo) [inline sky/core.py:771]: Move this check into `storage.py`?

File: sky/core.py [hunk 5]
```diff
@@ -757,3 +757,21 @@ def storage_delete(name: str) -> None:
...
+
+    # Add stores from tuple to storage object
+    if 'GCS' in stores:
```
Comment (michaelzhiluo) [inline sky/core.py:774]: Is it possible to do this in parallel? (i.e. I dont want to wait for the bucket to first upload to GCS and then S3 if I specify GCS,S3 for stores)

File: tests/test_smoke.py [hunk 6]
```diff
@@ -968,6 +968,22 @@ def test_new_bucket_creation_and_deletion(self, tmp_local_storage_obj,
         out = subprocess.check_output(['sky', 'storage', 'ls'])
         assert tmp_local_storage_obj.name not in out.decode('utf-8')
 
+    def test_sky_storage_cli(self, tmp_bucket_name, tmp_source):
+        # Creates new bucket with local source
+        out1 = subprocess.check_output(
+            ['sky', 'storage', 'create', tmp_bucket_name, tmp_source, 'AWS'])
```
Comment (michaelzhiluo) [inline tests/test_smoke.py:974]: `AWS` should be `S3`. Might want to double check if the test passes.

File: sky/cli.py [hunk 7]
```diff
@@ -2635,6 +2635,32 @@ def storage_delete(names: Tuple[str], all: bool):  # pylint: disable=redefined-b
...
+          source: str; File path where the data is initially stored. Can be a
+            local path or a cloud URI (s3://, gs://, etc.). Local paths do not
+            need to be absolute.
+          stores: Optional; Specify pre-initialized stores (S3 or GCS).
+
+    Examples:
+
+    .. code-block:: bash
+
+        # Create a storage object with GCS and AWS stores
```
Comment (michaelzhiluo) [inline sky/cli.py:2656]: Nit: `GCS and S3 stores`

**Review by michaelzhiluo — CHANGES_REQUESTED** [review]

File: sky/core.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 import sys
 from typing import Any, Dict, List, Optional, Tuple
 
-from sky import dag
+from sky import dag, StoreType
```
[the remark this answers is not in the document]
Comment (choochutran) [reply sky/core.py:7]: Added from sky.data import storage as storage_lib and changed to storage_lib.StoreType when referencing enum.

File: sky/cli.py (hunk 7 above)
Comment (choochutran) [reply sky/cli.py:2656]: Changed AWS to S3.

File: tests/test_smoke.py (hunk 6 above)
Comment (choochutran) [reply tests/test_smoke.py:974]: Changed to S3 in test and corresponding methods.

File: sky/cli.py [hunk 9]
```diff
@@ -2635,6 +2635,32 @@ def storage_delete(names: Tuple[str], all: bool):  # pylint: disable=redefined-b
         sky.storage_delete(name)
 
 
+@storage.command('create', cls=_DocumentedCodeCommand)
+@click.argument('name', required=False, type=str, nargs=1)
```
[the remark this answers is not in the document]
Comment (choochutran) [reply sky/cli.py:2639]: Made `name` a required argument.

File: sky/cli.py (hunk 1 above)
Comment (choochutran) [reply sky/cli.py:2641]: Added `default=['S3']`

File: sky/core.py (hunk 4 above)
Comment (choochutran) [reply sky/core.py:771]: The add_store method takes in a StoreType in `storage.py`. Should I make a separate function that calls add_store that takes in a list of strings as a parameter?

File: sky/cli.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (choochutran) [reply sky/cli.py:2657]: I set `nargs=1` for name and source so it takes those two arguments respectively and `nargs=-1` for the stores argument so click creates a tuple containing the rest of the arguments for stores.

[the remark this answers is not in the document]
Comment (choochutran) [reply sky/cli.py:2657]: Added more examples as well.

File: sky/cli.py [hunk 10]
```diff
@@ -2635,6 +2635,53 @@ def storage_delete(names: Tuple[str], all: bool):  # pylint: disable=redefined-b
...
+
+    Examples:
+
+    .. code-block:: bash
+
+        # Create an empty storage object named imagenet with GCS and S3 stores
+        sky storage create --name imagenet GCS S3
+
+        # Create a storage object from bucket URL using source bucket as the backing store
+        sky storage create --source S3://bucket-name
```
Comment (michaelzhiluo) [inline sky/cli.py:2672]: Make sure to run these commands before putting them as example^ Name is required and name is not specified here

File: sky/cli.py [hunk 11]
```diff
@@ -2635,6 +2635,53 @@ def storage_delete(names: Tuple[str], all: bool):  # pylint: disable=redefined-b
...
+
+        # Create storage object by uploading contents from local source path to
+            imagenet S3 bucket and creates new bucket if necessary
+        sky storage create --name imagenet --source /Users/skypilotUser/Project S3
```
Comment (michaelzhiluo) [inline sky/cli.py:2676]: Incosistent use of arg specifiers (--name, --source, but no --stores?)

File: sky/core.py [hunk 12]
```diff
@@ -757,3 +760,24 @@ def storage_delete(name: str) -> None:
                                     source=handle.source,
                                     sync_on_reconstruction=False)
         store_object.delete()
+
+
+def storage_create(name: Optional[str] = None,
+                   source: Optional[str] = None,
+                   stores: Tuple[str] = None):
+    """"Create a storage.
```
Comment (michaelzhiluo) [inline sky/core.py:768]: Nit: `Creates a storage object.`

File: tests/test_smoke.py [hunk 13]
```diff
@@ -968,6 +968,22 @@ def test_new_bucket_creation_and_deletion(self, tmp_local_storage_obj,
         out = subprocess.check_output(['sky', 'storage', 'ls'])
         assert tmp_local_storage_obj.name not in out.decode('utf-8')
```
Comment (michaelzhiluo) [inline tests/test_smoke.py:970]: Does this test fully pass? (double check)

**Comment by michaelzhiluo** [comment]
@choochutran How is the PR progress going?

File: sky/cli.py (hunk 2 above)
Comment (choochutran) [reply sky/cli.py]: Added help descriptions for all click arguments and options.

File: sky/core.py (hunk 4 above)
Comment (choochutran) [reply sky/core.py]: Handled alongside concurrency

File: sky/core.py (hunk 5 above)
Comment (choochutran) [reply sky/core.py]: Implemented concurrently adding stores by defining a top level helper function in core.py to resolve pickling errors due to passing in a storage object method to p.map.

File: sky/core.py (hunk 12 above)
Comment (choochutran) [reply sky/core.py]: Fixed

File: tests/test_smoke.py (hunk 13 above)
Comment (choochutran) [reply tests/test_smoke.py]: Test passes S3 case but my GCS project/credentials are buggy so might need validation on your end.

Comment (michaelzhiluo) [reply tests/test_smoke.py]: Try to fix the GCS credentials on your end.

File: docs/source/reference/storage.rst [hunk 14]
```diff
@@ -244,7 +244,21 @@ To manage persistent Storage objects, the sky CLI provides two useful commands -
...
     sky-dataset-romil  3 mins ago  S3     sky launch -c demo examples/storage_demo.yaml  READY
 
-2.  :code:`sky storage delete` allows you to delete any Storage objects managed
+2.  :code:`sky storage create` allows you to create a sky Storage object. Common use cases include:
+   - Create a storage object by uploading contents from local source path to specified stores
+        :code:`$ sky storage create imagenet -src /Users/skypilotUser/Project -s S3 -s GCS`
+   - Create an empty storage object with GCS stores
+        :code:`$ sky storage create emptyStore -s GCS`
+   - Create a storage object by uploading local source contents to default S3 store
+        :code:`$ sky storage create cleanedData /Users/skypilotUser/Project/CleanedData`
```
Comment (michaelzhiluo) [inline docs/source/reference/storage.rst:253]: `--source` needs to be specified?

File: docs/source/reference/storage.rst [hunk 15]
```diff
@@ -244,7 +244,21 @@ To manage persistent Storage objects, the sky CLI provides two useful commands -
...
+   - Create a storage object from bucket URL using source bucket to infer name and as the backing store
+        :code:`$ sky storage -src S3://bucket-name`
```
Comment (michaelzhiluo) [inline docs/source/reference/storage.rst:255]: Add code comment saying the bucket will be named `bucket-name`

File: docs/source/reference/storage.rst [hunk 16]
```diff
@@ -244,7 +244,21 @@ To manage persistent Storage objects, the sky CLI provides two useful commands -
...
+.. code-block:: console
+
+    $ sky storage create imagenet -src S3://bucket-name -s GCS -s AWS
```
Comment (michaelzhiluo) [inline docs/source/reference/storage.rst:258]: This would break right?

File: sky/cli.py [hunk 17]
```diff
@@ -1845,12 +1844,12 @@ def stop(
               help='Skip confirmation prompt.')
 @usage_lib.entrypoint
 def autostop(
-    clusters: Tuple[str],
-    all: Optional[bool],  # pylint: disable=redefined-builtin
-    idle_minutes: Optional[int],
-    cancel: bool,  # pylint: disable=redefined-outer-name
-    down: bool,  # pylint: disable=redefined-outer-name
-    yes: bool,
+        clusters: Tuple[str],
```
Comment (michaelzhiluo) [inline sky/cli.py:1847]: Make sure that you have the most recent version of dependencies for `format.sh`

File: sky/cli.py [hunk 18]
```diff
@@ -3403,21 +3451,21 @@ def bench():
...
-    gpus: Optional[str],
-    num_nodes: Optional[int],
-    use_spot: Optional[bool],
-    image_id: Optional[str],
-    env: List[Dict[str, str]],
-    disk_size: Optional[int],
-    idle_minutes_to_autostop: Optional[int],
-    yes: bool,
+        entrypoint: str,
+        benchmark: str,
```
Comment (michaelzhiluo) [inline sky/cli.py:3455]: Similar comment^

File: docs/source/reference/storage.rst (hunk 14 above)
Comment (choochutran) [reply docs/source/reference/storage.rst:253]: If source isn't specified, it creates an empty Sky Storage object through the requested cloud provider.

File: docs/source/reference/storage.rst (hunk 15 above)
Comment (choochutran) [reply docs/source/reference/storage.rst:255]: Can't be public bucket

Comment (michaelzhiluo) [reply docs/source/reference/storage.rst:255]: Mention in Line 254: `Create's a storage object of name `bucket-name`, which is inferred from the bucket's URL. The bucket must be private bucket.

File: docs/source/reference/storage.rst (hunk 16 above)
Comment (michaelzhiluo) [reply docs/source/reference/storage.rst:258]: `bucket-name` != `imagenet`

File: docs/source/reference/storage.rst (hunk 15 above)
Comment (choochutran) [reply docs/source/reference/storage.rst:255]: Mention that external public buckets are okay and that a user's publically accessible bucket is okay as well.

File: tests/test_smoke.py (hunk 13 above)
Comment (michaelzhiluo) [reply tests/test_smoke.py]: We discussed more on this, need to reset gcloud on local laptop.

File: docs/source/reference/storage.rst (hunk 16 above)
Comment (choochutran) [reply docs/source/reference/storage.rst]: Fixed

**Pull request closed without merging**